Source/Modules/javascript.cxx

```diff
@@ -459,22 +597,448 @@ int JAVASCRIPT::nativeWrapper(Node *n) {
+ * Note, that if the class is annotated with "feature:ignore" or "feature:interface"
+ * it tries to use next base class.
+ *
+ * @param n The node that represents the class we want the base class to
+ * @return the string representation of the base class name or null if the class
+ * does not extend other(s).
+ */
+String *JAVASCRIPT::getBaseClass(Node *n)
+{
+  String *returnBaseClassName = NULL;
+
+  String *typemapLookupCppType = Getattr(n, "name");
+  String *typescriptTypemapBaseClassName = typemapLookup(n, "typescriptbase", typemapLookupCppType);
+  if (typescriptTypemapBaseClassName && Strcmp(typescriptTypemapBaseClassName, "") != 0)
+  {
+    return typescriptTypemapBaseClassName;
+  }
+
+  List *baselist = Getattr(n, "bases");
+  if (!baselist)
+  {
+    return NULL;
+  }
+
+  Iterator base = First(baselist);
+  bool alreadyHaveBaseClass = false;
+  while (base.item)
+  {
+    if (!(GetFlag(base.item, "feature:ignore") || Getattr(base.item, "feature:interface")))
+    {
+      SwigType *baseclassname = Getattr(base.item, "name");
+      returnBaseClassName = baseclassname;
+
+      String *name = getProxyName(baseclassname);
+      if (name)
+      {
+        returnBaseClassName = name;
+      }
+
+      if (alreadyHaveBaseClass)
+      {
+        String *proxyclassname = Getattr(n, "classtypeobj");
+        Swig_warning(WARN_JAVA_MULTIPLE_INHERITANCE, Getfile(n), Getline(n),
+                     "Warning for %s, base %s ignored. Multiple inheritance is not supported in TypeScript.\n", SwigType_namestr(proxyclassname), SwigType_namestr(baseclassname));
+      }
+      alreadyHaveBaseClass = true;
+    }
+    base = Next(base);
+  }
+  return returnBaseClassName;
+}
+
+/**
+ * Get the name of a Type to be used as TypeScript interface/enum name
+ *
+ * 1 - Try to find a class corresponding to the node received as parameter.
+ * 2 - If the class does not exist try to find an enum
+ * 3 - If a class or an enum does not exist return NULL
+ * 4 - If the node found have already a "proxyname" attribute, return it
+ * 5 - Otherwise get the sym:name property from the class and return it
+ *
+ * @param T A node representing the Type to which is necessary to get the proxyname
+ *
+ * @return A string with the Type name to be used in the TypeScript proxy Interface
+ * or null if cannot find a class or enum or the node does not have "sym:name" property
+ */
+String *JAVASCRIPT::getProxyName(SwigType *t)
+{
+  String *proxyname = NULL;
+
+  Node *n = classLookup(t);
+  if (!n)
+  {
+    n = enumLookup(t);
+  }
+  if (!n)
+  {
+    return NULL;
+  }
+
+  proxyname = Getattr(n, "proxyname");
+  if (proxyname)
+  {
+    return proxyname;
+  }
+
+  String *symname = Getattr(n, "sym:name");
+  if (!symname)
+  {
+    return NULL;
+  }
+  proxyname = Copy(symname);
+  Setattr(n, "proxyname", proxyname); // Cache it
+  return proxyname;
+}
+
+/**
+ * Try to find a typemap related to a node
```

Comment:
This doesn't look JS/TS-specific... are you sure there is no existing function that could be used instead?